Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -2036,6 +2050,45 @@ public void UpdateStatus(bool overwriteable = true, int expectedCallbackFrequenc
                     Contract.Assert(m_performanceAggregator != null, "Adaptive IO requires non-null performanceAggregator");
                     m_pipQueue.AdjustIOParallelDegree(m_perfInfo);
                 }
+
+                if (m_scheduleConfiguration.EarlyWorkerRelease)
+                {
+                    PerformEarlyReleaseWorker(numProcessPipsPending, numProcessPipsRunning);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Decide whether we can release a remote worker. This method is executed every 2 seconds depending on the frequency of LogStatus timer.
+        /// </summary>
+        private void PerformEarlyReleaseWorker(long numProcessPipsPending, long numProcessPipsRunning)
+        {
+            long numProcessPipsWaiting = numProcessPipsPending - numProcessPipsRunning;
+
+            // If the available remote workers perform at half capacity in future, how many process pips we can concurrently execute:
+            int totalSlots = (int)Math.Ceiling(Workers.Where(a => a.IsRemote && a.IsAvailable).Sum(a => a.TotalProcessSlots * 0.5) + LocalWorker.TotalProcessSlots);

```

Comment:
>int totalSlots = (int)Math.Ceiling(Workers.Where(a => a.IsRemote && a.IsAvailable).Sum(a => a.TotalProcessSlots * 0.5) + LocalWorker.TotalProcessSlots); [](start = 12, length = 152)

I think the logic should be slightly different.

var workerToReleaseCandidate = Workers.Where(a => a.IsRemote && a.IsAvailable).OrderBy(a => a.AcquiredProcessSlots).FirstOrDefault();
if (workerToReleaseCandidate != null) return;

// NOTE: Multiplier is configurable.
int totalSlots = m_scheduleConfiguration.EarlyWorkerReleaseCapacityMultiplier * (int)Math.Ceiling(Workers.Where(a => a.IsRemote && a.IsAvailable).Sum(a => a.TotalProcessSlots) + LocalWorker.TotalProcessSlots);

// Release worker if numProcessPipsWaiting can be satisfied by remaining workers
if (numProcessPipsWaiting > 0 && (numProcessPipsWaiting < totalSlots - workerToReleaseCandidate.TotalProcessSlots))
{
     workerToReleaseCandidate.EarlyReleaseAsync();
}

Reply:
@semihokur - How does this algorithm compare to the one the QB uses?

Reply:
I have no idea what algorithm they uses.  This is what we discussed in the perf meeting though. This seems a good start and we can do several iterations for that logic. 

---
In reply to: [289506464](https://github.com/microsoft/BuildXL/pull/392#discussion_r289506464) [](ancestors = 289506464)